pyccel/parser/syntactic.py

```diff
@@ -633,54 +642,110 @@ def _visit_FunctionDef(self, stmt):
-                i = i+1
-
-            txt  = '#$ header ' + name
-            txt += '(' + ','.join(types) + ')'
+                        ls = ls if isinstance(ls, (PythonTuple)) else list(ls)
+
+                while i<len(ls) :
+                    arg = ls[i]
+                    if isinstance(arg, Symbol):
+                        arg = arg.name
+                        types.append(arg)
+                    elif isinstance(arg, String):
+                        arg = str(arg)
+                        arg = arg.strip("'").strip('"')
+                        types.append(arg)
+                    else:
+                        msg = 'Invalid argument type of type {} passed to template decorator'.format(type(arg))
+                        errors.report(msg,
+                                    symbol = comb_types,
+                                    bounding_box = (stmt.lineno, stmt.col_offset),
+                                    severity='error')
 
-            if results:
-                txt += ' results(' + ','.join(results) + ')'
+                    i += 1
+    
+                txt  = '#$ header template ' + str(tp_name)
+                txt += '(' + '|'.join(types) + ')'
+                templates += [hdr_parse(stmts=txt)]
 
-            header = hdr_parse(stmts=txt)
-            if name in self.namespace.static_functions:
-                header = header.to_static()
+        # extract the types to construct a header
+        if 'types' in decorators:
+            if not isinstance(decorators['types'], list):
+                decorators['types'] = [decorators['types']]
+            for comb_types in decorators['types']:
+                types = []
+                results = []
+                container = types
+                i = 0
+                ls = comb_types.args
+    
+                while i<len(ls) :
+                    arg = ls[i]
+    
+                    if isinstance(arg, Symbol):
+                        arg = arg.name
+                        container.append(arg)
+                    elif isinstance(arg, String):
+                        arg = str(arg)
+                        arg = arg.strip("'").strip('"')
+                        container.append(arg)
+                    elif isinstance(arg, ValuedArgument):
+                        arg_name = arg.name
+                        arg  = arg.value
+                        container = results
+                        if not arg_name == 'results':
+                            msg = 'Argument "{}" provided to the types decorator is not valid'.format(arg_name)
+                            errors.report(msg,
+                                        symbol = comb_types['types'],
+                                        bounding_box = (stmt.lineno, stmt.col_offset),
+                                        severity='error')
+                        else:
+                            ls = arg if isinstance(arg, PythonTuple) else [arg]
+                            i = -1
```

Comment:
Is this why you use a while? To recurse?

I think this would be more legible as a function

Reply:
Yes that's why it's used. you think I should make a function for the while block, it's only used in one place

Reply:
This only works because the results have to be given by keyword and the arguments have to be given by position right? If the order was changed then you might miss some values

Reply:
I'm happy to leave it as it is for now and get a second opinion from other reviewers when the branch is passing about what is more legible

Reply:
I also find this hard to read, because the index `i` starts from `0` and is either incremented by `1` or set to `-1`...

Reply:
I will try to change it